dspinner/pin.go

```diff
@@ -0,0 +1,996 @@
+// Package dspinner implements structures and methods to keep track of
+// which objects a user wants to keep stored locally.  This implementation
+// stores pin data in a datastore.
+package dspinner
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"path"
+	"sync"
+
+	"github.com/ipfs/go-cid"
+	ds "github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	ipfspinner "github.com/ipfs/go-ipfs-pinner"
+	"github.com/ipfs/go-ipfs-pinner/dsindex"
+	ipld "github.com/ipfs/go-ipld-format"
+	logging "github.com/ipfs/go-log"
+	mdag "github.com/ipfs/go-merkledag"
+	"github.com/ipfs/go-merkledag/dagutils"
+	"github.com/polydawn/refmt/cbor"
+)
+
+const (
+	basePath     = "/pins"
+	pinKeyPath   = "/pins/pin"
+	indexKeyPath = "/pins/index"
+	dirtyKeyPath = "/pins/state/dirty"
+)
+
+var (
+	// ErrNotPinned is returned when trying to unpin items that are not pinned.
+	ErrNotPinned = fmt.Errorf("not pinned or pinned indirectly")
+
+	log logging.StandardLogger = logging.Logger("pin")
+
+	linkDirect, linkRecursive string
+
+	pinCidDIndexPath string
+	pinCidRIndexPath string
+	pinNameIndexPath string
+
+	dirtyKey = ds.NewKey(dirtyKeyPath)
+)
+
+func init() {
+	directStr, ok := ipfspinner.ModeToString(ipfspinner.Direct)
+	if !ok {
+		panic("could not find Direct pin enum")
+	}
+	linkDirect = directStr
+
+	recursiveStr, ok := ipfspinner.ModeToString(ipfspinner.Recursive)
+	if !ok {
+		panic("could not find Recursive pin enum")
+	}
+	linkRecursive = recursiveStr
+
+	pinCidRIndexPath = path.Join(indexKeyPath, "cidRindex")
+	pinCidDIndexPath = path.Join(indexKeyPath, "cidDindex")
+	pinNameIndexPath = path.Join(indexKeyPath, "nameIndex")
+}
+
+// pinner implements the Pinner interface
+type pinner struct {
+	lock sync.RWMutex
+
+	dserv  ipld.DAGService
+	dstore ds.Datastore
+
+	cidDIndex dsindex.Indexer
+	cidRIndex dsindex.Indexer
+	nameIndex dsindex.Indexer
+
+	clean int64
+	dirty int64
+}
+
+var _ ipfspinner.Pinner = (*pinner)(nil)
+
+type pin struct {
+	id       string
+	cid      cid.Cid
+	metadata map[string]interface{}
+	mode     ipfspinner.Mode
+	name     string
+}
+
+func (p *pin) dsKey() ds.Key {
+	return ds.NewKey(path.Join(pinKeyPath, p.id))
+}
+
+func newPin(c cid.Cid, mode ipfspinner.Mode, name string) *pin {
+	return &pin{
+		id:   ds.RandomKey().String(),
+		cid:  c,
+		name: name,
+		mode: mode,
+	}
+}
+
+type syncDAGService interface {
+	ipld.DAGService
+	Sync() error
+}
+
+// New creates a new pinner and loads its keysets from the given datastore. If
+// there is no data present in the datastore, then an empty pinner is returned.
+func New(ctx context.Context, dstore ds.Datastore, dserv ipld.DAGService) (ipfspinner.Pinner, error) {
```

Comment:
This function doesn't actually respect the context (i.e. neither of the expensive operations like `loadAllPins` or `rebuildIndexes` actually abort when the context is cancelled). If you're able to fix this up in a non-painful way so that we can associates timeouts with `New` like we do with the IpldPinnner then that would be great.

However, I'm ok treating this as a TODO we cover in a followup PR.

Reply:
Done.

Checking context when loading pinner and during iterative operations.